Migrate multi_margin_loss to ATen (CUDA)

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61426 Migrate multi_margin_loss to ATen (CUDA)**

Closes gh-24600, closes gh-24601

These operators use custom kernels that aren't well suited to `TensorIterator` style, so this is just changing the CPU code and cleaning up the style.

Differential Revision: [D29648015](https://our.internmc.facebook.com/intern/diff/D29648015)

peterbell10 committed:
Migrate multi_margin_loss to ATen (CUDA)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/MultiMarginLoss.cu:
```diff
@@ -0,0 +1,388 @@
+        }
+        at::sum_out(out, tmp_output, /*dims=*/IntArrayRef{});
+      }
+    } else {
```
can this ever happen? input.ndim() <= 2 is checked in the multi_margin_loss_shape_check.

peterbell10 replied:
Yeah, I think you're right. I've changed it to an internal assert in the 2d case.

ngimel reviewed (commented):
This looks good, I have very minor comments

peterbell10 committed:
Update on "Migrate multi_margin_loss to ATen (CUDA)"


Closes gh-24600, closes gh-24601

These operators use custom kernels that aren't well suited to `TensorIterator` style, so this is just changing the CPU code and cleaning up the style.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/MultiMarginLoss.cu:
```diff
@@ -0,0 +1,384 @@
+      }
+    } else {
+      auto in_sizes = input.sizes();
+      TORCH_INTERAL_ASSERT(in_sizes.size() == 2);
```
typo here, hence real build errors.

peterbell10 committed:
Update on "Migrate multi_margin_loss to ATen (CUDA)"


Closes gh-24600, closes gh-24601

These operators use custom kernels that aren't well suited to `TensorIterator` style, so this is just changing the CPU code and cleaning up the style.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29648015).
